libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -653,50 +654,40 @@ internal object MapboxRouteLineUtils {
-                    // causes problems in creating the traffic expression since the expression
-                    // values need to be in strictly ascending order. A value of 0 can be caused
-                    // by the first point in a route leg being the same as the last point in the
-                    // previous route leg. There may be other causes as well.
-                    if (distance > 0.0) {
-                        val percentDistanceTraveled = runningDistance / route.distance()
-                        val isInAClosure = closureRanges.any { it.contains(index) }
-                        val congestionValue: String = when {
-                            isInAClosure -> RouteLayerConstants.CLOSURE_CONGESTION_VALUE
-                            trafficCongestion.isNullOrEmpty() ->
-                                RouteLayerConstants.UNKNOWN_CONGESTION_VALUE
-                            index >= trafficCongestion.size ->
-                                RouteLayerConstants.UNKNOWN_CONGESTION_VALUE
-                            else -> trafficCongestion[index]
-                        }
-                        val roadClass = getRoadClassForIndex(roadClassArray, index)
-
-                        itemsToReturn.add(
-                            ExtractedRouteData(
-                                runningDistance,
-                                percentDistanceTraveled,
-                                congestionValue,
-                                roadClass,
-                                legIndex,
-                                isLegOrigin
-                            )
-                        )
-                        isLegOrigin = false
-                        runningDistance += distance
+
+                granularDistances.legsDistances[legIndex].forEachIndexed { index, distanceObject ->
+                    val isInAClosure = closureRanges.any { it.contains(index) }
+                    val congestionValue: String = when {
+                        isInAClosure -> RouteLayerConstants.CLOSURE_CONGESTION_VALUE
+                        trafficCongestion.isNullOrEmpty() ->
+                            RouteLayerConstants.UNKNOWN_CONGESTION_VALUE
+                        index >= trafficCongestion.size ->
+                            RouteLayerConstants.UNKNOWN_CONGESTION_VALUE
+                        else -> trafficCongestion[index]
                     }
+                    val roadClass = getRoadClassForIndex(roadClassArray, index)
+
+                    val offset =
+                        1.0 - distanceObject.distanceRemaining / granularDistances.completeDistance
+                    itemsToReturn.add(
+                        ExtractedRouteData(
+                            offset,
+                            congestionValue,
+                            roadClass,
+                            legIndex,
+                            index == 0
```

Comment:
Why is leg index hardcoded to 0? This is used for styling the active leg differently than non-active legs.

Reply:
`legIndex` is not hardcoded here, it comes from the list we're iterating over. The `index == 0` produces a boolean value depending on which geometry index in a leg we are, if it's the first index, we know it's the origin of the leg. I added a named parameter to make this clear.